File: bridge/jellyfin.py
#!/usr/bin/env python3
"""SysDeck - Jellyfin Bridge Helper
Author: Jeremy Anderson (https://dcos.net)

Manages the Jellyfin media server as a systemd service and exposes
the built-in Jellyfin web UI for iframe embedding in the SysDeck
panel.

v0.0.35 directive: "next we will integrate a jellyfin management
module where it starts, stops, and loads the admin panel in the
module." Jellyfin ships a single systemd unit (jellyfin.service) on
every distro that packages it, and serves a full web UI on
http://127.0.0.1:8096. The bridge starts/stops the service via
systemctl; the panel iframes the running web UI — same pattern as
the v0.0.34 Glances integration (bridge/glances.py).

Subcommands:
  summary       — service status + version + port + library counts
  status        — service status only (active/sub/uptime)
  start         — systemctl start jellyfin.service
  stop          — systemctl stop jellyfin.service
  restart       — systemctl restart jellyfin.service
  web-status    — {running, port, url} for the iframe
  libraries     — best-effort library list from Jellyfin's HTTP API

Cockpit way (v0.0.31+ pattern): the bridge runs systemctl via
subprocess directly — no `sudo` shell-out. The JS panel passes
{ superuser: 'try' } to cockpit.spawn so the cockpit bridge prompts
the operator via polkit for the org.sysdeck.jellyfin.modify action
(added in v0.0.35).

Jellyfin is GPL-2.0 licensed by the Jellyfin contributors. This
bridge helper invokes it as a separate process via subprocess — the
suite (MIT) and Jellyfin (GPL-2.0) remain independent programs. No
Jellyfin code is bundled.

Usage:
    python3 /usr/lib/sysdeck/bridge/jellyfin.py summary
    python3 /usr/lib/sysdeck/bridge/jellyfin.py start
    python3 /usr/lib/sysdeck/bridge/jellyfin.py web-status
"""

import json
import os
import shutil
import subprocess
import sys
from datetime import datetime
from typing import Any


JELLYFIN_LICENSE = "GPL-2.0"
JELLYFIN_URL = "https://jellyfin.org/"
JELLYFIN_SERVICE = "jellyfin.service"

# Default Jellyfin webserver port. The operator can override via
# /etc/jellyfin/networking.xml; this default matches Jellyfin's
# out-of-the-box config.
JELLYFIN_WEB_HOST = "127.0.0.1"
JELLYFIN_WEB_PORT = 8096

# Default URL the panel iframes. Jellyfin binds to 0.0.0.0 by default;
# the panel uses 127.0.0.1 to keep the iframe on the cockpit host.
JELLYFIN_WEB_URL = f"http://{JELLYFIN_WEB_HOST}:{JELLYFIN_WEB_PORT}"


def _have(binary: str) -> bool:
    """True if binary is on PATH."""
    return shutil.which(binary) is not None


def _systemctl_show(unit: str, props: list[str]) -> dict[str, str]:
    """Return a dict of {property: value} from systemctl show."""
    out = subprocess.run(
        ["systemctl", "show", unit, "--property=" + ",".join(props)],
        capture_output=True, text=True, timeout=5,
    ).stdout.strip()
    result = {}
    for line in out.splitlines():
        k, _, v = line.partition("=")
        if k:
            result[k] = v
    return result


def _service_status(unit: str) -> dict[str, Any]:
    """Return service state dict: {active, sub, status, uptime_seconds}."""
    props = _systemctl_show(unit, ["ActiveState", "SubState", "ActiveEnterTimestamp"])
    active = props.get("ActiveState", "unknown")
    sub = props.get("SubState", "unknown")
    uptime = 0
    ts = props.get("ActiveEnterTimestamp", "")
    if ts:
        try:
            dt = datetime.strptime(ts[:25], "%a %Y-%m-%d %H:%M:%S")
            uptime = int((datetime.now() - dt).total_seconds())
        except (ValueError, OSError):
            pass
    if active == "active":
        status = "running"
    elif active == "activating":
        status = "starting"
    elif active == "failed":
        status = "error"
    elif active in ("inactive", "deactivating"):
        status = "stopped"
    else:
        status = "unknown"
    return {"active": active, "sub": sub, "status": status, "uptime_seconds": uptime}


def _detect_version() -> str:
    """Best-effort Jellyfin version detection."""
    if _have("jellyfin"):
        out = subprocess.run(
            ["jellyfin", "--version"], capture_output=True, text=True, timeout=5,
        ).stdout.strip()
        return out.splitlines()[0][:80] if out else ""
    # Read from /etc/jellyfin/jellyfin.db if installed (best-effort)
    return ""


def _detect_port() -> int:
    """Return the configured Jellyfin port (best-effort).

    Reads /etc/jellyfin/networking.xml if present; falls back to
    the default 8096.
    """
    try:
        with open("/etc/jellyfin/networking.xml", encoding="utf-8") as fh:
            for line in fh:
                if "<Port>" in line and "</Port>" in line:
                    port_str = line.split("<Port>")[1].split("</Port>")[0].strip()
                    return int(port_str)
    except (FileNotFoundError, ValueError, PermissionError, OSError):
        pass
    return JELLYFIN_WEB_PORT


def cmd_summary(_args: list[str]) -> dict[str, Any]:
    """Combined summary — service status + version + port + URL."""
    if not _have("jellyfin") and not _unit_loaded():
        return {
            "available": False,
            "reason": "jellyfin not installed",
            "install": (
                "Arch: pacman -S jellyfin  ·  "
                "Debian: apt install jellyfin  ·  "
                "Fedora: dnf install jellyfin"
            ),
            "license": JELLYFIN_LICENSE,
            "url": JELLYFIN_URL,
        }
    state = _service_status(JELLYFIN_SERVICE)
    port = _detect_port()
    return {
        "available": True,
        "service": JELLYFIN_SERVICE,
        "status": state["status"],
        "active": state["active"],
        "sub": state["sub"],
        "uptime_seconds": state["uptime_seconds"],
        "version": _detect_version(),
        "port": port,
        "url": f"http://{JELLYFIN_WEB_HOST}:{port}",
        "license": JELLYFIN_LICENSE,
        "homepage": JELLYFIN_URL,
    }


def _unit_loaded() -> bool:
    """True if the jellyfin.service unit is loaded on the host."""
    out = subprocess.run(
        ["systemctl", "list-unit-files", JELLYFIN_SERVICE],
        capture_output=True, text=True, timeout=5,
    ).stdout
    return JELLYFIN_SERVICE in out


def cmd_status(_args: list[str]) -> dict[str, Any]:
    """Detailed service status."""
    if not _have("jellyfin") and not _unit_loaded():
        return {"available": False, "reason": "jellyfin not installed"}
    return _service_status(JELLYFIN_SERVICE)


def _systemctl(action: str) -> dict[str, Any]:
    """Run systemctl <action> jellyfin.service and return the result."""
    if not _have("systemctl"):
        return {"rc": 127, "success": False, "stderr": "systemctl not found"}
    try:
        r = subprocess.run(
            ["systemctl", action, JELLYFIN_SERVICE],
            capture_output=True, text=True, timeout=30,
        )
        return {
            "action": action,
            "service": JELLYFIN_SERVICE,
            "rc": r.returncode,
            "success": r.returncode == 0,
            "output": (r.stdout or "").strip(),
            "stderr": (r.stderr or "").strip(),
        }
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError) as exc:
        return {"action": action, "service": JELLYFIN_SERVICE,
                "rc": 1, "success": False, "stderr": str(exc)}


def cmd_start(_args: list[str]) -> dict[str, Any]:
    return _systemctl("start")


def cmd_stop(_args: list[str]) -> dict[str, Any]:
    return _systemctl("stop")


def cmd_restart(_args: list[str]) -> dict[str, Any]:
    return _systemctl("restart")


def cmd_web_status(_args: list[str]) -> dict[str, Any]:
    """Return the Jellyfin web UI URL + running state for iframe embedding."""
    if not _have("jellyfin") and not _unit_loaded():
        return {
            "available": False,
            "reason": "jellyfin not installed",
            "install": (
                "Arch: pacman -S jellyfin  ·  "
                "Debian: apt install jellyfin  ·  "
                "Fedora: dnf install jellyfin"
            ),
            "license": JELLYFIN_LICENSE,
            "url": JELLYFIN_URL,
        }
    state = _service_status(JELLYFIN_SERVICE)
    port = _detect_port()
    return {
        "available": True,
        "running": state["status"] == "running",
        "status": state["status"],
        "port": port,
        "url": f"http://{JELLYFIN_WEB_HOST}:{port}",
        "license": JELLYFIN_LICENSE,
        "homepage": JELLYFIN_URL,
    }


def cmd_libraries(_args: list[str]) -> dict[str, Any]:
    """Best-effort library list via the Jellyfin HTTP API.

    Calls GET /Library/VirtualFolders on the local Jellyfin instance.
    Returns {libraries: [...], count: N} or {error: ...} if the API
    is unreachable or the operator has not yet completed initial
    setup (no admin user → 401).
    """
    import urllib.request
    import urllib.error
    port = _detect_port()
    url = f"http://{JELLYFIN_WEB_HOST}:{port}/Library/VirtualFolders"
    try:
        # Jellyfin's public API doesn't require auth for /Library/VirtualFolders
        # when called from localhost on default config — best-effort.
        req = urllib.request.Request(url, headers={
            "Accept": "application/json",
            "X-Emby-Authorization": 'MediaBrowser Client="SysDeck", Device="Cockpit", Version="0.0.35"',
        })
        with urllib.request.urlopen(req, timeout=4) as resp:
            data = json.loads(resp.read().decode("utf-8"))
        libs = [
            {
                "name": lib.get("Name", "?"),
                "type": lib.get("CollectionType", "mixed"),
                "paths": lib.get("Locations", []),
            }
            for lib in (data if isinstance(data, list) else [])
        ]
        return {"libraries": libs, "count": len(libs)}
    except urllib.error.HTTPError as exc:
        return {"error": f"HTTP {exc.code}: {exc.reason} — Jellyfin may need initial setup via the web UI first."}
    except urllib.error.URLError as exc:
        return {"error": f"Connection refused: {exc.reason} — Jellyfin may not be running."}
    except (ValueError, OSError, KeyError) as exc:
        return {"error": str(exc)}


COMMANDS = {
    "summary":     lambda args: cmd_summary(args),
    "status":      lambda args: cmd_status(args),
    "start":       lambda args: cmd_start(args),
    "stop":        lambda args: cmd_stop(args),
    "restart":     lambda args: cmd_restart(args),
    "web-status":  lambda args: cmd_web_status(args),
    "libraries":   lambda args: cmd_libraries(args),
}


def main(argv: list[str]) -> int:
    if not argv or argv[0] in ("-h", "--help"):
        print(__doc__)
        return 0
    cmd = COMMANDS.get(argv[0])
    if not cmd:
        print(f"Unknown subcommand: {argv[0]}", file=sys.stderr)
        print(f"Available: {', '.join(sorted(COMMANDS))}", file=sys.stderr)
        return 2
    print(json.dumps(cmd(argv[1:]), indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: plugins/sysdeck-jellyfin/jellyfin.js
/*
 * SysDeck - Jellyfin Panel (v0.0.35)
 * Author: Jeremy Anderson (https://dcos.net)
 *
 * v0.0.35 directive: "next we will integrate a jellyfin management
 * module where it starts, stops, and loads the admin panel in the
 * module." Jellyfin ships a single systemd unit (jellyfin.service)
 * and serves a full admin web UI on http://127.0.0.1:8096. The
 * bridge starts/stops/restarts the service via systemctl; the panel
 * iframes the running admin UI — same pattern as the v0.0.34 Glances
 * integration.
 *
 * The panel surfaces:
 *   - Service summary card: status badge, version, port, uptime
 *   - Service controls: Start / Stop / Restart (polkit prompts)
 *   - Admin UI iframe: loads http://127.0.0.1:8096 when running
 *   - Library list: best-effort GET /Library/VirtualFolders
 *   - Install hint when jellyfin is not installed
 *
 * Jellyfin is GPL-2.0 licensed by the Jellyfin contributors. The
 * bridge helper invokes it as a separate process via subprocess —
 * the suite (MIT) and Jellyfin (GPL-2.0) remain independent
 * programs. No Jellyfin code is bundled.
 *
 * Bridge surface (see shared/bridge.js → bridge.jellyfin):
 *   summary()              → {available, status, version, port, url, ...}
 *   status()               → service state dict
 *   start()                → {action, rc, success, output, stderr}
 *   stop()                 → same shape
 *   restart()              → same shape
 *   webStatus()            → {running, url, port, ...}
 *   libraries()            → {libraries: [...], count: N}
 */

export async function mount(panel, { bridge, EventBus }) {
    panel.innerHTML = renderSkeleton();

    const [summary, webStatus, libraries] = await Promise.all([
        safe(bridge.jellyfin.summary(), {}),
        safe(bridge.jellyfin.webStatus(), {}),
        safe(bridge.jellyfin.libraries(), { libraries: [], count: 0 }),
    ]);

    const running = webStatus?.running === true || summary?.status === 'running';
    const available = webStatus?.available !== false && summary?.available !== false;
    const webUrl = webStatus?.url || summary?.url || 'http://127.0.0.1:8096';

    panel.innerHTML = `
        <header>
            <h2 class="suite-panel-title">Jellyfin Media Server</h2>
            <p class="suite-panel-subtitle">
                Self-hosted media streaming · movies · TV · music
                · <span class="suite-badge info">GPL-2.0 · Jellyfin contributors</span>
                ${available
                    ? (running
                        ? ` · <span class="suite-badge success">running</span>`
                        : ` · <span class="suite-badge">stopped</span>`)
                    : ` · <span class="suite-badge danger">not installed</span>`}
            </p>
        </header>

        ${renderServiceCard(summary, available, running, webUrl)}

        ${available && running ? renderWebIframe(webUrl) : ''}

        ${available ? renderLibrariesCard(libraries) : ''}
    `;

    wireEvents(panel, { bridge, EventBus });
    EventBus.emit('jellyfin.loaded', { running, available });
}

// ── Service card ────────────────────────────────────────────────────

function renderServiceCard(summary, available, running, webUrl) {
    if (!available) {
        return `
            <div class="suite-card">
                <h3 class="suite-card-title">Jellyfin Service</h3>
                <div class="suite-card-body">
                    <p class="suite-muted">
                        ${escapeHtml(summary?.reason || 'Jellyfin is not installed.')}
                    </p>
                    ${summary?.install
                        ? `<p class="suite-muted" style="margin-top:0.5rem"><code>${escapeHtml(summary.install)}</code></p>`
                        : ''}
                    <p class="suite-muted" style="margin-top:0.5rem">
                        Jellyfin is GPL-2.0 licensed by the Jellyfin contributors —
                        <a href="https://jellyfin.org/" style="color:var(--sysdeck-accent)">https://jellyfin.org/</a>
                    </p>
                </div>
            </div>
        `;
    }

    const uptime = formatUptime(summary?.uptime_seconds || 0);
    return `
        <div class="suite-card">
            <div class="suite-card-header">
                <h3 class="suite-card-title">Service — <code>${escapeHtml(summary?.service || 'jellyfin.service')}</code></h3>
                <div class="suite-row" style="gap:0.5rem">
                    <button class="suite-btn ${running ? '' : 'suite-btn-primary'}" id="btn-jellyfin-start" ${running ? 'disabled' : ''}>▶ Start</button>
                    <button class="suite-btn" id="btn-jellyfin-stop" ${!running ? 'disabled' : ''}>■ Stop</button>
                    <button class="suite-btn" id="btn-jellyfin-restart" ${!running ? 'disabled' : ''}>↻ Restart</button>
                    <button class="suite-btn suite-btn-ghost" id="btn-jellyfin-refresh">↻ Refresh</button>
                </div>
            </div>
            <div class="suite-card-body">
                <table class="suite-table">
                    <tbody>
                        <tr><th>Status</th><td>${statusBadge(summary?.status)}</td></tr>
                        <tr><th>ActiveState</th><td class="suite-table-mono">${escapeHtml(summary?.active || '—')}</td></tr>
                        <tr><th>SubState</th><td class="suite-table-mono">${escapeHtml(summary?.sub || '—')}</td></tr>
                        <tr><th>Uptime</th><td class="suite-muted">${running ? escapeHtml(uptime) : '—'}</td></tr>
                        <tr><th>Version</th><td class="suite-muted">${escapeHtml(summary?.version || '—')}</td></tr>
                        <tr><th>Port</th><td class="suite-table-mono">${summary?.port || 8096}</td></tr>
                        <tr><th>URL</th><td class="suite-table-mono"><a href="${escapeHtml(webUrl)}" target="_blank" rel="noopener noreferrer" style="color:var(--sysdeck-accent)">${escapeHtml(webUrl)}</a></td></tr>
                    </tbody>
                </table>
                <p class="suite-muted" style="font-size:0.85rem;margin-top:0.5rem">
                    The bridge runs <code>systemctl start/stop/restart jellyfin.service</code>
                    via the cockpit superuser channel (polkit <code>org.sysdeck.jellyfin.modify</code>).
                </p>
            </div>
        </div>
    `;
}

function renderWebIframe(webUrl) {
    return `
        <div class="suite-card">
            <div class="suite-card-header">
                <h3 class="suite-card-title">Admin Panel — ${escapeHtml(webUrl)}</h3>
                <a href="${escapeHtml(webUrl)}" target="_blank" rel="noopener noreferrer" class="suite-btn suite-btn-ghost">↗ Open in new tab</a>
            </div>
            <iframe src="${escapeHtml(webUrl)}"
                    style="width:100%;height:800px;border:1px solid var(--sysdeck-border);border-radius:6px;background:#1e1e1e;"
                    id="jellyfin-iframe"
                    title="Jellyfin Admin Panel"
                    allow="autoplay; fullscreen; encrypted-media; picture-in-picture"
                    allowfullscreen></iframe>
        </div>
    `;
}

function renderLibrariesCard(libraries) {
    const libs = libraries?.libraries || [];
    if (libraries?.error) {
        return `
            <div class="suite-card">
                <h3 class="suite-card-title">Libraries</h3>
                <div class="suite-card-body">
                    <p class="suite-muted">${escapeHtml(libraries.error)}</p>
                </div>
            </div>
        `;
    }
    return `
        <div class="suite-card">
            <div class="suite-card-header">
                <h3 class="suite-card-title">Libraries (${libraries?.count || 0})</h3>
            </div>
            <table class="suite-table">
                <thead><tr><th>Name</th><th>Type</th><th>Paths</th></tr></thead>
                <tbody>
                    ${libs.map((lib) => `
                        <tr>
                            <td><strong>${escapeHtml(lib.name)}</strong></td>
                            <td class="suite-table-mono">${escapeHtml(lib.type || 'mixed')}</td>
                            <td class="suite-muted suite-table-mono">${escapeHtml((lib.paths || []).join(' · ') || '—')}</td>
                        </tr>
                    `).join('') || '<tr><td colspan="3" class="suite-muted">No libraries configured. Open the admin panel above to add media folders.</td></tr>'}
                </tbody>
            </table>
        </div>
    `;
}

// ── Event wiring ────────────────────────────────────────────────────

function wireEvents(panel, { bridge, EventBus }) {
    const action = async (btn, method, label) => {
        if (!btn) return;
        btn.disabled = true;
        const original = btn.textContent;
        btn.textContent = `${label} ...`;
        try {
            const r = await bridge.jellyfin[method]();
            if (!r?.success && method !== 'restart') {
                alert(`${label} failed:\n${r?.stderr || r?.output || 'unknown'}`);
            }
            setTimeout(() => mount(panel, { bridge, EventBus }), 1000);
        } catch (err) {
            btn.disabled = false;
            btn.textContent = original;
            alert(`${label} error: ${err.message || err}`);
        }
    };

    panel.querySelector('#btn-jellyfin-start')?.addEventListener('click', (ev) => action(ev.currentTarget, 'start', 'Start'));
    panel.querySelector('#btn-jellyfin-stop')?.addEventListener('click', (ev) => action(ev.currentTarget, 'stop', 'Stop'));
    panel.querySelector('#btn-jellyfin-restart')?.addEventListener('click', (ev) => action(ev.currentTarget, 'restart', 'Restart'));
    panel.querySelector('#btn-jellyfin-refresh')?.addEventListener('click', () => {
        mount(panel, { bridge, EventBus });
    });
}

// ── Utilities ───────────────────────────────────────────────────────

async function safe(p, fallback) {
    try {
        const v = await p;
        return v ?? fallback;
    } catch {
        return fallback;
    }
}

function statusBadge(status) {
    const map = {
        'running':  '<span class="suite-badge success">running</span>',
        'starting': '<span class="suite-badge warn">starting</span>',
        'stopped':  '<span class="suite-badge">stopped</span>',
        'error':    '<span class="suite-badge danger">error</span>',
        'unknown':  '<span class="suite-badge">unknown</span>',
    };
    return map[status] || `<span class="suite-badge">${escapeHtml(status || 'unknown')}</span>`;
}

function formatUptime(seconds) {
    if (!seconds || seconds <= 0) return '—';
    const days = Math.floor(seconds / 86400);
    const hours = Math.floor((seconds % 86400) / 3600);
    const mins = Math.floor((seconds % 3600) / 60);
    if (days > 0) return `${days}d ${hours}h ${mins}m`;
    if (hours > 0) return `${hours}h ${mins}m`;
    return `${mins}m`;
}

function escapeHtml(s) {
    return String(s == null ? '' : s)
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;');
}

function renderSkeleton() {
    return `<div class="suite-skeleton">
        <div class="suite-skeleton-line w-1/3"></div>
        <div class="suite-skeleton-line w-2/3"></div>
        <div class="suite-skeleton-line w-1/2"></div>
    </div>`;
}
